### PR TITLE
Add launcher shortcuts on Linux

### PR DESCRIPTION
**Pull request by spacey-sooty**
Doesn't add desktop shortcuts due to complications with how gnome
handles desktop files

Resolves https://github.com/wpilibsuite/WPILibInstaller-Avalonia/issues/136

### CODE REVIEW

**Comment by sciencewhiz** [comment]
Using a category of Robotics instead of Development might be more relevant

**Comment by calcmogul** [comment]
Do custom categories work?

**Comment by calcmogul** [comment]
We probably don't want a .desktop for ToolsUpdater.
```
[tav@myriad applications]$ ls
'AdvantageScope (WPILib) 2025.desktop'  kscreen-doctor.desktop                  'SysId 2025.desktop'
'Choreo (WPILib) 2025.desktop'          mimeinfo.cache                          'ToolsUpdater 2025.desktop'
'DataLogTool 2025.desktop'              'OutlineViewer 2025.desktop'            wine
'FRC VS Code 2025.desktop'              'PathWeaver 2025.desktop'               wine-extension-chm.desktop
'Glass 2025.desktop'                    'roboRIOTeamNumberSetter 2025.desktop'  wine-extension-hlp.desktop
kscreen-doctor-2.desktop                'RobotBuilder 2025.desktop'             wine-extension-msp.desktop
kscreen-doctor-3.desktop                'Shuffleboard 2025.desktop'             wine-extension-vbs.desktop
kscreen-doctor-4.desktop                'SmartDashboard 2025.desktop'
```

**Comment by sciencewhiz** [comment]
> Do custom categories like Robotics work?

It's in the spec. https://specifications.freedesktop.org/menu-spec/latest/additional-category-registry.html

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 1]
```diff
@@ -31,6 +31,52 @@ public class InstallPageViewModel : PageViewModelBase
...
+Version=1.0
+Type=Application
+Categories=Development
+Name={name} {frcYear}
+Comment={nameNoWPILib} tool for the 2025 FIRST Robotics Competition season
+Exec={configurationProvider.InstallDirectory}/tools/{nameNoWPILib}.sh
+Icon={configurationProvider.InstallDirectory}/frccode/wpilib-256.ico
+Terminal=false
+StartupNotify=true
+StartupWMClass=Code
```
Comment (sciencewhiz) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:52]: This needs to be different per tool. See https://ubuntuhandbook.org/index.php/2024/04/missing-icon-dock-ubuntu-2404/ for how to determine what it should be.

**Review by sciencewhiz — COMMENTED** [review]
I just tried this on Ubuntu 24.04, and I didn't get tool files installed to ~/local/share/applications, just VS code

**Comment by spacey-sooty** [comment]
> I just tried this on Ubuntu 24.04, and I didn't get tool files installed to ~/local/share/applications, just VS code

Was it a tools only run?

**Comment by sciencewhiz** [comment]
> > I just tried this on Ubuntu 24.04, and I didn't get tool files installed to ~/local/share/applications, just VS code
> 
> Was it a tools only run? Assuming it was its fixed now.

Yes it was tools only, and it's fixed now

**Comment by sciencewhiz** [comment]
It seems like most of the tools don't show up in the dock when I ran them

**Comment by spacey-sooty** [comment]
> It seems like most of the tools don't show up in the dock when I ran them

For me tho showed up after a bit of a wait. I think I might've had to interact with the window as well?

**Comment by spacey-sooty** [comment]
[Screencast from 2024-10-26 23-03-13.webm](https://github.com/user-attachments/assets/498c5f26-2648-4bb4-b73e-a28048080d0c)

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 2]
```diff
@@ -911,29 +959,42 @@ [Desktop Entry]
...
+                }
+
+                CreateLinuxShortcut("Choreo (WPILib)", frcYear, "Choreo", token);
+                CreateLinuxShortcut("AdvantageScope (WPILib)", frcYear, "AdvantageScope", token);
+                CreateLinuxShortcut("Glass", frcYear, "Glass - DISCONNECTED", token);
+                CreateLinuxShortcut("OutlineViewer", frcYear, "OutlineViewer - DISCONNECTED", token);
+                CreateLinuxShortcut("DataLogTool", frcYear, "Datalog Tool", token);
+                CreateLinuxShortcut("SysId", frcYear, "System Identification", token);
+                CreateLinuxShortcut("SmartDashboard", frcYear, "SmartDashboard | Disconnected", token);
+                CreateLinuxShortcut("RobotBuilder", frcYear, "FRC RobotBuilder", token);
```
Comment (sciencewhiz) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:994]:
```suggestion
                CreateLinuxShortcut("RobotBuilder", frcYear, "robotbuilder-RobotBuilder", token);

```

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 3]
```diff
@@ -911,29 +959,42 @@ [Desktop Entry]
...
+                CreateLinuxShortcut("PathWeaver", frcYear, "PathWeaver - 2025.1.1-beta-1", token);
+                CreateLinuxShortcut("roboRIOTeamNumberSetter", frcYear, "roboRIO Team Number Setter", token);
+                CreateLinuxShortcut("Shuffleboard", frcYear, "Shuffleboard", token);
```
Comment (sciencewhiz) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:997]:
```suggestion
                CreateLinuxShortcut("Shuffleboard", frcYear, "edu.wpi.first.shuffleboard.app.Shuffleboard", token);

```

**Comment by sciencewhiz** [comment]
There isn't a delay if launching either the executable or shell script directly, only when using the launcher

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 4]
```diff
@@ -911,29 +959,42 @@ [Desktop Entry]
...
+                    }, token);
+                }
+
+                CreateLinuxShortcut("Choreo (WPILib)", frcYear, "Choreo", token);
+                CreateLinuxShortcut("AdvantageScope (WPILib)", frcYear, "AdvantageScope", token);
+                CreateLinuxShortcut("Glass", frcYear, "Glass - DISCONNECTED", token);
+                CreateLinuxShortcut("OutlineViewer", frcYear, "OutlineViewer - DISCONNECTED", token);
+                CreateLinuxShortcut("DataLogTool", frcYear, "Datalog Tool", token);
+                CreateLinuxShortcut("SysId", frcYear, "System Identification", token);
+                CreateLinuxShortcut("SmartDashboard", frcYear, "SmartDashboard | Disconnected", token);
```
Comment (sciencewhiz) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:993]:
```suggestion
                CreateLinuxShortcut("SmartDashboard", frcYear, "edu-wpi-first-smartdashboard-SmartDashboard", token);

```

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 5]
```diff
@@ -911,29 +959,42 @@ [Desktop Entry]
...
+                CreateLinuxShortcut("RobotBuilder", frcYear, "FRC RobotBuilder", token);
+                CreateLinuxShortcut("PathWeaver", frcYear, "PathWeaver - 2025.1.1-beta-1", token);
```
Comment (sciencewhiz) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:995]:
```suggestion
                CreateLinuxShortcut("PathWeaver", frcYear, "edu.wpi.first.pathweaver.PathWeaver", token);

```

**Review by sciencewhiz — COMMENTED** [review]
The java tools StartupWMClass weren't correct

**Comment by sciencewhiz** [comment]
The .desktop files are written with CRLF line endings

**Comment by spacey-sooty** [comment]
> The .desktop files are written with CRLF line endings

idrk how we can fix this, the dotnet formatter will want everything in the file to be crlf

**Comment by spacey-sooty** [comment]
I dont think it being written with CRLF endings is an issue and even if it is its seperate to this PR, the vscode desktop file already had this problem.

**Comment by calcmogul** [comment]
These shortcuts aren't showing up in a Robotics category for me.
Screenshot_20241111_185532
Screenshot_20241111_185408

**Comment by calcmogul** [comment]
This still didn't sort correctly. The following seems to work though.
```diff
diff --git a/WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs b/WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs
index 989b108..b9a1744 100644
--- a/WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs
+++ b/WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs
@@ -42,7 +42,7 @@ namespace WPILibInstaller.ViewModels
 [Desktop Entry]
 Version=1.0
 Type=Application
-Categories=Science & Maths
+Categories=Robotics;Science
 Name={name} {frcYear}
 Comment={nameNoWPILib} tool for the 2025 FIRST Robotics Competition season
 Exec={configurationProvider.InstallDirectory}/tools/{nameNoWPILib}.sh
@@ -59,7 +59,7 @@ StartupWMClass={wmClass}
 [Desktop Entry]
 Version=1.0
 Type=Application
-Categories=Robotics
+Categories=Robotics;Science
 Name={name} {frcYear}
 Comment={name} tool for the 2025 FIRST Robotics Competition season
 Exec={configurationProvider.InstallDirectory}/tools/{name}.sh
```

**Review by calcmogul — APPROVED** [review]

**Comment by sciencewhiz** [comment]
Needs a shortcut for elastic

**Comment by sciencewhiz** [comment]
On Ubuntu 24.04, only a few are showing up in the launcher (Advantagescope, Elastic, VS Code)

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 6]
```diff
@@ -31,6 +31,52 @@ public class InstallPageViewModel : PageViewModelBase
...
+";
+
+            }
+            else
+            {
+                contents = $@"#!/usr/bin/env xdg-open
+[Desktop Entry]
+Version=1.0
+Type=Application
+CategoriesRobotics;Science
```
Comment (sciencewhiz) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:62]:
```suggestion
Categories=Robotics;Science

```

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 7]
```diff
@@ -31,6 +31,53 @@ public class InstallPageViewModel : PageViewModelBase
...
+
+            }
+            else
+            {
+                contents = $@"#!/usr/bin/env xdg-open
+[Desktop Entry]
+Version=1.0
+Type=Application
+Categories=Robotics;Science
+
```
Comment (calcmogul) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:63]: Why was this blank line added in the bugfix? The other .desktop doesn't have it.

Comment (spacey-sooty) [reply WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:63]: Mistake

Comment (spacey-sooty) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:63]:
```suggestion
```

**Review by sciencewhiz — APPROVED** [review]

**Pull request merged**